File: C#/Geometry/GeometryCalculator.cs
/*
 * Geometry Calculator
 * Author: Андрій Будильников
 * 
 * This program calculates geometric properties of various shapes including
 * areas, perimeters, volumes, and surface areas.
 */

using System;

namespace Geometry
{
    /// <summary>
    /// Provides methods for calculating geometric properties of various shapes
    /// </summary>
    public class GeometryCalculator
    {
        // Mathematical constant Pi
        private const double PI = 3.14159265359;
        
        /// <summary>
        /// Calculates the area of a circle
        /// </summary>
        /// <param name="radius">Radius of the circle</param>
        /// <returns>Area of the circle</returns>
        public static double CircleArea(double radius)
        {
            if (radius < 0)
            {
                throw new ArgumentException("Radius cannot be negative");
            }
            
            return PI * radius * radius;
        }
        
        /// <summary>
        /// Calculates the circumference of a circle
        /// </summary>
        /// <param name="radius">Radius of the circle</param>
        /// <returns>Circumference of the circle</returns>
        public static double CircleCircumference(double radius)
        {
            if (radius < 0)
            {
                throw new ArgumentException("Radius cannot be negative");
            }
            
            return 2 * PI * radius;
        }
        
        /// <summary>
        /// Calculates the area of a triangle using Heron's formula
        /// </summary>
        /// <param name="sideA">Length of side A</param>
        /// <param name="sideB">Length of side B</param>
        /// <param name="sideC">Length of side C</param>
        /// <returns>Area of the triangle</returns>
        public static double TriangleArea(double sideA, double sideB, double sideC)
        {
            if (sideA <= 0 || sideB <= 0 || sideC <= 0)
            {
                throw new ArgumentException("All sides must be positive");
            }
            
            // Check triangle inequality
            if (sideA + sideB <= sideC || sideA + sideC <= sideB || sideB + sideC <= sideA)
            {
                throw new ArgumentException("Invalid triangle: sides do not satisfy triangle inequality");
            }
            
            // Heron's formula
            double semiPerimeter = (sideA + sideB + sideC) / 2;
            return Math.Sqrt(semiPerimeter * (semiPerimeter - sideA) * (semiPerimeter - sideB) * (semiPerimeter - sideC));
        }
        
        /// <summary>
        /// Calculates the perimeter of a triangle
        /// </summary>
        /// <param name="sideA">Length of side A</param>
        /// <param name="sideB">Length of side B</param>
        /// <param name="sideC">Length of side C</param>
        /// <returns>Perimeter of the triangle</returns>
        public static double TrianglePerimeter(double sideA, double sideB, double sideC)
        {
            if (sideA <= 0 || sideB <= 0 || sideC <= 0)
            {
                throw new ArgumentException("All sides must be positive");
            }
            
            return sideA + sideB + sideC;
        }
        
        /// <summary>
        /// Calculates the area of a rectangle
        /// </summary>
        /// <param name="length">Length of the rectangle</param>
        /// <param name="width">Width of the rectangle</param>
        /// <returns>Area of the rectangle</returns>
        public static double RectangleArea(double length, double width)
        {
            if (length < 0 || width < 0)
            {
                throw new ArgumentException("Length and width cannot be negative");
            }
            
            return length * width;
        }
        
        /// <summary>
        /// Calculates the perimeter of a rectangle
        /// </summary>
        /// <param name="length">Length of the rectangle</param>
        /// <param name="width">Width of the rectangle</param>
        /// <returns>Perimeter of the rectangle</returns>
        public static double RectanglePerimeter(double length, double width)
        {
            if (length < 0 || width < 0)
            {
                throw new ArgumentException("Length and width cannot be negative");
            }
            
            return 2 * (length + width);
        }
        
        /// <summary>
        /// Calculates the volume of a sphere
        /// </summary>
        /// <param name="radius">Radius of the sphere</param>
        /// <returns>Volume of the sphere</returns>
        public static double SphereVolume(double radius)
        {
            if (radius < 0)
            {
                throw new ArgumentException("Radius cannot be negative");
            }
            
            return (4.0 / 3.0) * PI * Math.Pow(radius, 3);
        }
        
        /// <summary>
        /// Calculates the surface area of a sphere
        /// </summary>
        /// <param name="radius">Radius of the sphere</param>
        /// <returns>Surface area of the sphere</returns>
        public static double SphereSurfaceArea(double radius)
        {
            if (radius < 0)
            {
                throw new ArgumentException("Radius cannot be negative");
            }
            
            return 4 * PI * radius * radius;
        }
        
        /// <summary>
        /// Main method to demonstrate geometry calculations
        /// </summary>
        public static void Main(string[] args)
        {
            Console.WriteLine("Geometry Calculator Demo");
            Console.WriteLine("======================");
            
            // Circle calculations
            double circleRadius = 5.0;
            Console.WriteLine($"Circle with radius {circleRadius}:");
            Console.WriteLine($"  Area: {CircleArea(circleRadius):F2}");
            Console.WriteLine($"  Circumference: {CircleCircumference(circleRadius):F2}");
            
            // Triangle calculations
            double sideA = 3.0, sideB = 4.0, sideC = 5.0;
            Console.WriteLine($"\nTriangle with sides {sideA}, {sideB}, {sideC}:");
            Console.WriteLine($"  Area: {TriangleArea(sideA, sideB, sideC):F2}");
            Console.WriteLine($"  Perimeter: {TrianglePerimeter(sideA, sideB, sideC):F2}");
            
            // Rectangle calculations
            double length = 6.0, width = 4.0;
            Console.WriteLine($"\nRectangle with length {length} and width {width}:");
            Console.WriteLine($"  Area: {RectangleArea(length, width):F2}");
            Console.WriteLine($"  Perimeter: {RectanglePerimeter(length, width):F2}");
            
            // Sphere calculations
            double sphereRadius = 3.0;
            Console.WriteLine($"\nSphere with radius {sphereRadius}:");
            Console.WriteLine($"  Volume: {SphereVolume(sphereRadius):F2}");
            Console.WriteLine($"  Surface Area: {SphereSurfaceArea(sphereRadius):F2}");
        }
    }
}

File: Python/Geometry/geometry_calculator.py
"""
Geometry Calculator
Author: Андрій Будильников

This program calculates geometric properties of various shapes including
areas, perimeters, volumes, and surface areas.
"""

import math

class GeometryCalculator:
    """
    Provides methods for calculating geometric properties of various shapes
    """
    
    """Mathematical constant Pi"""
    PI = 3.14159265359
    
    @staticmethod
    def circle_area(radius):
        """
        Calculates the area of a circle
        :param radius: Radius of the circle
        :return: Area of the circle
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        
        return GeometryCalculator.PI * radius * radius
    
    @staticmethod
    def circle_circumference(radius):
        """
        Calculates the circumference of a circle
        :param radius: Radius of the circle
        :return: Circumference of the circle
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        
        return 2 * GeometryCalculator.PI * radius
    
    @staticmethod
    def triangle_area(side_a, side_b, side_c):
        """
        Calculates the area of a triangle using Heron's formula
        :param side_a: Length of side A
        :param side_b: Length of side B
        :param side_c: Length of side C
        :return: Area of the triangle
        """
        if side_a <= 0 or side_b <= 0 or side_c <= 0:
            raise ValueError("All sides must be positive")
        
        """Check triangle inequality"""
        if (side_a + side_b <= side_c or 
            side_a + side_c <= side_b or 
            side_b + side_c <= side_a):
            raise ValueError("Invalid triangle: sides do not satisfy triangle inequality")
        
        """Heron's formula"""
        semi_perimeter = (side_a + side_b + side_c) / 2
        return math.sqrt(semi_perimeter * (semi_perimeter - side_a) * 
                        (semi_perimeter - side_b) * (semi_perimeter - side_c))
    
    @staticmethod
    def triangle_perimeter(side_a, side_b, side_c):
        """
        Calculates the perimeter of a triangle
        :param side_a: Length of side A
        :param side_b: Length of side B
        :param side_c: Length of side C
        :return: Perimeter of the triangle
        """
        if side_a <= 0 or side_b <= 0 or side_c <= 0:
            raise ValueError("All sides must be positive")
        
        return side_a + side_b + side_c
    
    @staticmethod
    def rectangle_area(length, width):
        """
        Calculates the area of a rectangle
        :param length: Length of the rectangle
        :param width: Width of the rectangle
        :return: Area of the rectangle
        """
        if length < 0 or width < 0:
            raise ValueError("Length and width cannot be negative")
        
        return length * width
    
    @staticmethod
    def rectangle_perimeter(length, width):
        """
        Calculates the perimeter of a rectangle
        :param length: Length of the rectangle
        :param width: Width of the rectangle
        :return: Perimeter of the rectangle
        """
        if length < 0 or width < 0:
            raise ValueError("Length and width cannot be negative")
        
        return 2 * (length + width)
    
    @staticmethod
    def sphere_volume(radius):
        """
        Calculates the volume of a sphere
        :param radius: Radius of the sphere
        :return: Volume of the sphere
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        
        return (4.0 / 3.0) * GeometryCalculator.PI * (radius ** 3)
    
    @staticmethod
    def sphere_surface_area(radius):
        """
        Calculates the surface area of a sphere
        :param radius: Radius of the sphere
        :return: Surface area of the sphere
        """
        if radius < 0:
            raise ValueError("Radius cannot be negative")
        
        return 4 * GeometryCalculator.PI * radius * radius

def main():
    """
    Main function to demonstrate geometry calculations
    """
    print("Geometry Calculator Demo")
    print("======================")
    
    """Circle calculations"""
    circle_radius = 5.0
    print(f"Circle with radius {circle_radius}:")
    print(f"  Area: {GeometryCalculator.circle_area(circle_radius)}")
    print(f"  Circumference: {GeometryCalculator.circle_circumference(circle_radius)}")
    
    """Triangle calculations"""
    side_a, side_b, side_c = 3.0, 4.0, 5.0
    print(f"\nTriangle with sides {side_a}, {side_b}, {side_c}:")
    print(f"  Area: {GeometryCalculator.triangle_area(side_a, side_b, side_c)}")
    print(f"  Perimeter: {GeometryCalculator.triangle_perimeter(side_a, side_b, side_c)}")
    
    """Rectangle calculations"""
    length, width = 6.0, 4.0
    print(f"\nRectangle with length {length} and width {width}:")
    print(f"  Area: {GeometryCalculator.rectangle_area(length, width)}")
    print(f"  Perimeter: {GeometryCalculator.rectangle_perimeter(length, width)}")
    
    """Sphere calculations"""
    sphere_radius = 3.0
    print(f"\nSphere with radius {sphere_radius}:")
    print(f"  Volume: {GeometryCalculator.sphere_volume(sphere_radius)}")
    print(f"  Surface Area: {GeometryCalculator.sphere_surface_area(sphere_radius)}")

if __name__ == "__main__":
    main()